app/controllers/companies_controller.rb

```diff
@@ -67,12 +67,30 @@ def create
 
     @company = Company.new(sanitize_params(company_params))
 
-    if @company.save
-      redirect_to @company, notice: t('.success')
+    saved = @company.save
+
+    unless request.xhr?
+      if @company.save
+        redirect_to @company, notice: t('.success')
+      else
+        flash.now[:alert] = t('.error')
+        render :new
+      end
+      return
+    end
+
+    # XHR request from modal in ShfApplication create view (to create company)
+    if saved
+      status = 'success'
+      id = 'company-number-entry'
+      html = helpers.company_number_entry_field(@company.company_number)
     else
-      flash.now[:alert] = t('.error')
-      render :new
+      status = 'errors'
+      id = 'company-create-errors'
+      html = helpers.model_errors_helper(@company)
     end
+
+    render json: { status: status, id: id, html: html }
```

Comment:
please use respond_to to make code more clear (you will not need so many ifs then)
https://apidock.com/rails/ActionController/MimeResponds/InstanceMethods/respond_to

Reply:
@zlekki - that is a good point but I'd like tot defer this restructuring until after this PR is merged.  

I'd also like to discuss the best way to do this with you at that time.